src/variable_info.hpp

```diff
@@ -16,98 +16,119 @@
+	impl::variable_info_state<vit> state_;
 	bool valid_;
+
+	void throw_on_invalid() const;
 	void calculate_value();
 };
 
-/// Extends variable_info with methods that can only be applied if vit != vit_const
-template<const variable_info_detail::variable_info_type vit>
-class non_const_variable_info : public variable_info<vit>, variable_info_detail::enable_if_non_const<vit>::type
+/**
+ * Additional functionality for a non-const variable_info.
+ * @todo: should these functions take a reference?
+ */
+template<typename V>
+class variable_info_mutable : public variable_info<V>
 {
 public:
-	non_const_variable_info(const std::string& name, config& game_vars) : variable_info<vit>(name, game_vars) {}
-	~non_const_variable_info() {}
+	variable_info_mutable(const std::string& name, config& game_vars)
+		: variable_info<V>(name, game_vars)
+	{
+		static_assert(!std::is_same<impl::vi_policy_const, typename std::remove_const<V>::type>::value,
+			"variable_info_mutable cannot be specialized with 'vi_const'");
+	}
 
-	/// clears the vale this object points to
-	/// if only_tables = true it will not clear attribute values.
-	/// might throw invalid_variablename_exception
-	void clear(bool only_tables = false) const;
+	/**
+	 * @returns                 The new appended range.
+	 * @throws                  invalid_variablename_exception
+	 */
+	config::child_itors append_array(std::vector<config> children) const;
+
+	/**
+	 * @returns                 The new inserted range.
+	 * @throws                  invalid_variablename_exception
+	 */
+	config::child_itors insert_array(std::vector<config> children) const;
+
+	/**
+	 * @returns                 The new range.
+	 * @throws                  invalid_variablename_exception
+	 */
+	config::child_itors replace_array(std::vector<config> children) const;
+
+	/**
+	 * @returns                 Merges.
+	 * @throws                  invalid_variablename_exception
+	 */
+	void merge_array(std::vector<config> children) const;
 
-	// the following 4 functions are used by [set_variables]
-	// they destroy the passed vector. (make it empty).
-
-	/// @return: the new appended range
-	/// might throw invalid_variablename_exception
-	config::child_itors append_array(std::vector<config> childs) const;
-	/// @return: the new inserted range
-	/// might throw invalid_variablename_exception
-	config::child_itors insert_array(std::vector<config> childs) const;
-	/// @return: the new range
-	/// might throw invalid_variablename_exception
-	config::child_itors replace_array(std::vector<config> childs) const;
-	/// merges
-	/// might throw invalid_variablename_exception
-	void merge_array(std::vector<config> childs) const;
+	/**
+	 * Clears the value this object points to.
+	 *
+	 * @param only_tables       If true, will not clear attribute values.
+	 * @throws                  invalid_variablename_exception
+	 */
+	void clear(bool only_tables = false) const;
 };
 
+/** See @ref variable_info_implementation::vi_policy_create. */
```

Comment:
vi_policy_create is an implementationd detail since its defeined in variable_info_detail.hpp you shouldn't redirect to detail header in the interface header. (people shouldn't need to understand the detail header in order ot use this class )